lib/irb/ruby-lex.rb

```diff
@@ -382,46 +384,69 @@ def calc_nesting_depth(opens)
-    line_index = lines.size - 2
-    if free_indent_token(opens, line_index)
-      return [indent, lines[line_index][/^ */].length].max
-    end
+  def free_indent_token?(token)
+    return false unless token
 
-    indent
+    FREE_INDENT_TOKENS.include?(token.event)
   end
 
-  def check_corresponding_token_depth(tokens, lines, line_index)
+  def process_indent_level(tokens, lines, line_index, is_newline)
     line_results = IRB::NestingParser.parse_by_line(tokens)
     result = line_results[line_index]
-    return unless result
+    if result
+      _tokens, prev_opens, next_opens, min_depth = result
+    else
+      # When last line is empty
+      prev_opens = next_opens = line_results.last[2]
+      min_depth = next_opens.size
+    end
 
     # To correctly indent line like `end.map do`, we use shortest open tokens on each line for indent calculation.
     # Shortest open tokens can be calculated by `opens.take(min_depth)`
-    _tokens, prev_opens, opens, min_depth = result
-    indent_level, _nesting_level = calc_nesting_depth(opens.take(min_depth))
-    indent = indent_level * 2
-    free_indent_tok = free_indent_token(opens, line_index)
-    prev_line_free_indent_tok = free_indent_token(prev_opens, line_index - 1)
-    if prev_line_free_indent_tok && prev_line_free_indent_tok != free_indent_tok
-      return indent
-    elsif free_indent_tok
-      return lines[line_index][/^ */].length
+    indent_level, _nesting_level = calc_nesting_depth(prev_opens.take(min_depth))
+    indent = 2 * indent_level
+
+    preserve_indent = lines[line_index - (is_newline ? 1 : 0)][/^ */].size
+
+    if free_indent_token?(prev_opens.last)
+      if is_newline && prev_opens.last.pos[0] == line_index
+        # First newline inside free-indent token
+        indent
+      else
+        # Accept any number of indent inside free-indent token
+        preserve_indent
+      end
+    elsif prev_opens.last&.event == :on_embdoc_beg || next_opens.last&.event == :on_embdoc_beg
+      if prev_opens.last&.event == next_opens.last&.event
+        # Accept any number of indent inside embdoc content
+        preserve_indent
+      else
+        # =begin or =end
+        0
+      end
+    elsif prev_opens.last&.event == :on_heredoc_beg
+      tok = prev_opens.last.tok
+      if prev_opens.size < next_opens.size || prev_opens.last == next_opens.last
+        if is_newline && lines[line_index].empty? && line_results[line_index - 1][1].last != prev_opens.last
+          # First line in heredoc
+          indent
+        elsif tok.match?(/^<<~/)
+          # Accept extra indent spaces inside `<<~` heredoc
+          [indent, preserve_indent].max
+        else
+          # Accept any number of indent inside other heredoc
+          preserve_indent
+        end
+      else
+        # Heredoc close
+        prev_line_indent_level, _prev_line_nesting_level = calc_nesting_depth(prev_opens)
+        tok.match?(/^<<[~-]/) ? 2 * (prev_line_indent_level - 1) : 0
```

Comment:
Just to make sure I understand it correctly: even in the heredoc closing token we can still access the full heredoc content, including the opening syntax?

Reply:
We can access only the opening syntax.

For this code
```
-| code          | prev_opens      | next_opens
1| if true       | []              | [if]
2|   <<A + <<~B  | [if]            | [if, <<~B, <<A]
3| content_a     | [if, <<~B, <<A] | [if, <<~B, <<A]
4| A             | [if, <<~B, <<A] | [if, <<~B]
5|     content_b | [if, <<~B]      | [if, <<~B]
6|   B           | [if, <<~B]      | [if]
7| end           | [if]            | []
```
In heredoc closing line (L4 and L6), `prev_opens.last` (`<<A` for line 4, `<<~B` for line 6) is the corresponding heredoc_beg token.

Reply:
Hmm, that's an very interesting behaviour. Though I guess that's required so the parser can later know what'd close the heredoc.
Thanks for explaining 👍